### DISCUSSION TITLE
Issue getting GenMon working with Powerzone Sync controller

### DISCUSSION THREAD

**Discussion by Dicion** [jgyates/genmon#1202] [General]
Hello!

I have followed all the instructions in the wiki, changed all the settings to match, and cannot communicate to the powerzone sync controller. This is on a brand new, just installed 500KW natural gas generator.

Please forgive the 'pictures of a monitor', as the pi is on a closed network that screenshots can't be easily taken off of.

I have verified that port 502 is open and can be connected to on the generator via nmap:

![image](https://github.com/user-attachments/assets/ad6db74d-9a90-4564-9262-9d176130b227)

I can access the generator controller via web interface on the ip address:

![image](https://github.com/user-attachments/assets/71c7815b-2ec8-465d-a946-b994e2ebc93b)

Here are the serial over tcp/ip settings as specified in the wiki:
(Controller type is set to powerzone on advanced page as well, didn't take a screenshot though)

![image](https://github.com/user-attachments/assets/653c46c6-7f79-4e0b-ac41-d726b7925008)

And here is the modbus log file showing nothing but errors:

![image](https://github.com/user-attachments/assets/843101c2-0241-4ce0-a10b-6e6a612c13d3)

I will gladly pull any other logs or data, get any additional screenshots, or try any changed settings on the Pi as needed. I CANNOT easily change settings on the generator as it is password locked by the vendor, and every time I need to make a change, I need to either call and convince them to give me a one time password, or have them come on-site to make the change.

I already read the Serial Troubleshooting and Tips and Tricks wiki pages. Neither have anything that seems to apply here since this is Modbus over TCP/IP.

Thank you.

**Comment by jgyates** [comment]
the good news is that you are getting responses back from the controller. The modbus excpetions in the log indicate that the controlelr is sending modbus exception packets back to genmon when a modbus register is requested. The exception code 0a explination can be be found in [this link](https://www.simplymodbus.ca/exceptions.htm). From the link:

"Gateway Path Unavailable - Specialized use in conjunction with gateways, indicates that the gateway was unable to allocate an internal communication path from the input port to the output port for processing the request. Usually means the gateway is misconfigured or overloaded."

I am guessing this has something to do with the TCP/IP gateway setting on the controller. Is the controller communicating with anything else over TCP/IP? If not you could set the IP gate way to the address of the pi and see if that works, but that is just a guess. I know you can not change the settings on the controller easily, but based on the data you present it looks like the only way to get this working is to change the settings on the controller. The IP port and address on genmon appear to be functioning because genmon is receiving a response (modbus exception packets) from the controller. Typically the industrial controllers send exception packets when they are not configured properly or you request a modbus register that is not implemented. Modbus defines 10 different exception codes.

**Comment by jgyates** [comment]
do you have the option of trying modbus RTU (serial)

https://github.com/jgyates/genmon/wiki/Appendix-G-Generac-H-100,-G-Panel-and-PowerZone-Controllers#powerzone-prosync

**Comment by Dicion** [comment]
> I am guessing this has something to do with the TCP/IP gateway setting on the controller. Is the controller communicating with anything else over TCP/IP? If not you could set the IP gate way to the address of the pi 

They both are already configured to the same IP address. Both the Pi and the Generator are on the same subnet, 10.11.5.0/24, and the gateway on both is set to 10.11.5.1. There should be no need to even use the gateway as it shouldn't need to even leave the subnet to get to the Pi.

Let me see if I can get the gateway IP changed to that of the Pi, or disable/remove it completely.

> do you have the option of trying modbus RTU (serial)

Yes, but again, this would require reconfiguration of the Generator again to enable. I will enable this as well at the same time as the above changes.


This may have to wait until the next time the Vendor comes out to the site, which should be this friday, as they are showing up to observe the first automated scheduled start/load bank test/cooldown/etc. I'll make the changes at that time, and see if it results in a different outcome.

**Comment by Dicion** [comment]
Doing additional research on this error yields the following:

"you're probably talking to a Modbus Gateway and you've misconfigured the Unit ID / Modbus Address to be a Unit ID that the gateway doesn't know about."

Gateway here doesn't mean tcp/IP gateway, but rather modbus gateway device, eg, the other end of the tcp connection.

What does it mean by unit ID? I'm not well versed in modbus myself. Is this a changeable value on genmon?

**Comment by jgyates** [comment]
Then you likely need to change the modbus address. I genmon you can change this on the advanced settings page. Double click the icon in the upper right of the web interface to get to the advanced settings page. I think the default for generac is 100 decimal or 64 hexadecimal (genmon expects the number to be in hex). Just make sure the address on the generator and genmon match.

**Comment by Dicion** [comment]
Okay thanks. I'll try this tomorrow and see how it goes.

**Comment by Dicion** [comment]
Update: Short version - Got it working!

Long version;

I tried 64 hex and it didn't work, so I was 100% prepared to start at 1 and go through FF one at a time until I found what it was set at. Turns out.. it was factory set at "1". So that was nice. You may wish to add that to the wiki the 'Newer Powerzone Sync controllers have been discovered set from the factory as modbus slave id 1, try that if the default, or 64 doesn't work'.

Also, the screenshots you have of the powerzone sync/pro controller screens look similar, but quite different than mine. I think it's likely that I have a much newer firmware load that those ones, which may also explain the '1' modbus setting. As I said above, this is a brand new unit just recently installed, and it has a total of like 5 hours on it as of right now :)

Additionally, I'm still seeing errors in the modbus log, but it seems to be working otherwise. I figured I'd post them for you in case you think it's relevant or needs to be looked into:

![image](https://github.com/user-attachments/assets/7d4b09f3-9265-427f-a1ff-e7cebaec386c)

Thank you for all your help, it certainly pointed me in the correct direction to get this solved!

**Reply by jgyates** [reply]
I am glad it is working for you. If you can send the logs from the About page. This will allow me to look at the logs and the registers.

**Reply by Dicion** [reply]
> I am glad it is working for you. If you can send the logs from the About page. This will allow me to look at the logs and the registers.

I submitted registers and logs via the email method from the about page about 5 minutes ago to you. If you didn't get them, let me know.

We just did a test run on a load bank in the last half hour, and Genmon was showing 4801 volts on it's reporting page. I think we're a decimal point off with this firmware or something. Hopefully you can deciper it.

However, we did get the notices that it was starting, running, errors, and stops, and that's the important part. The specific data is a nice to have but whatever :)

**Reply by Dicion** [reply]
Here's an example of the 4800 volts :)

![image](https://github.com/user-attachments/assets/293c4efb-00de-4162-a832-9612d98186a2)

**Reply by jgyates** [reply]
I got your registers and logs. It looks like the registers you submitted do not reflect the Voltage A-B, B-C and C-A readings above. Is that correct. The registers you submitted have all zeros for these values. Also, I am assuming the nominal value of these three registers should be 480V? Is that correct.

**Reply by Dicion** [reply]
Yes, sorry, I submitted the registers and such after the test run was done and the generator was back in standby mode, not while it was active.. I don't know if you can go back about half an hour before the submission and see data or not. 

And yes, it should be 480.

**Reply by Dicion** [reply]
If you need a submission while it is running, I can do it next Friday when we do our next test run. Just let me know.

**Reply by jgyates** [reply]
I checked in an update that should fix this for you. This change should correct Voltage A-B, B-C, C-A and Average Voltage readings. Let me know how it goes with your testing.

**Reply by Dicion** [reply]
Thank you, I will deploy the update and see how everything looks during our test run next week. 

Thank you for all your help.

**Reply by Dicion** [reply]
I sent you another set of registers and logs with it running and loaded on a 250A (Per phase) load bank.

Looks like the amperage is also off by a factor of 10, and the average current was Zero for whatever reason, but the voltage is correct now.

![chrome_oO3eiVtQ9T](https://github.com/user-attachments/assets/3556a6a2-5870-407e-90ab-dedc98fddddc)

**Reply by jgyates** [reply]
I will make this change. Looks like none of the current modbus registers are sowing anything excpet the Phase Current A,B and C. 

Also, I assume the power output should be 205.901 kW not 205901kW? correct?

**Reply by Dicion** [reply]
That is correct!

Thanks for all the help!

**Reply by jgyates** [reply]
I checked in an update that should fix these issues. I put a divided by 10 on the Average current but I left it in place. If it is like the H-Panel systems I think it may depend on how the system was setup (both inputs to the controller and controller settings) to determine what is output from the controller via modbus.

Let me know if you have any issues.

**Reply by Dicion** [reply]
I know it's been a few weeks, but I wanted to give you a final update that almost everything seems working properly with the latest changes you made. Average current is showing Zero, but everything else is perfect.

![image](https://github.com/user-attachments/assets/bda9eec5-240f-4160-8d13-11de19197fd1)

![image](https://github.com/user-attachments/assets/bdcd49d5-cb33-4017-99d3-9237802e43f4)

Thanks for all the fixes!

**Reply by jgyates** [reply]
Great, thanks for the update
